File: java_refactoring/src/main/java/org/example2/solution/HerbalTeaService.java
package org.example2.solution;

public class HerbalTeaService extends TeaService {
    public HerbalTeaService(int weight) {
        super(weight);
    }

    @Override
    protected TeaType getTeaType() {
        return TeaType.HERBAL;
    }

    @Override
    protected void makeTeaType() {
        System.out.println("Giving herbal tea");
        addAmountMade(getWeight() * .12 - .02);
        if (getTeaStrength() > .2 && getTeaFlavor() < 0) {
            System.out.println("Adding extra herbal tea strength to herbal tea");
            addTeaStrength(-9.2);
        }
    }

    @Override
    protected void pourTeaType() {
        System.out.println("Herbal tea is pouring");
        System.out.println("Herbal teas have been used since ancient times in Egypt, Greece, and China for medicinal and ritual purposes, long before true tea was widespread.");
        addAmountMade(-1.1);
    }
}

File: typescript_refactoring/example2/solution/HerbalTeaService.ts
import { TeaService } from "./TeaService";
import { TeaType } from "./TeaTypes";

export class HerbalTeaService extends TeaService {
    protected readonly teaType = TeaType.Herbal;

    public constructor(weight: number) {
        super(weight);
    }

    protected makeTeaType(): void {
        console.log("Giving herbal tea");
        this.amountMade += this.weight * 0.12 - 0.02;
        if (this.teaStrength > 0.2 && this.teaFlavor < 0) {
            console.log("Adding extra herbal tea strength to herbal tea");
            this.teaStrength += -9.2;
        }
    }

    protected pourTeaType(): void {
        console.log("Herbal tea is pouring");
        console.log("Herbal teas have been used since ancient times in Egypt, Greece, and China for medicinal and ritual purposes, long before true tea was widespread.");
        this.amountMade -= 1.1;
    }
}
